runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_base.c

```diff
@@ -264,3 +264,105 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_4x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_8x8x1_arm_64)
+
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x16x2_to_4x16x2_arm_64(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 4 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 2));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+  int32x4_t acc[16];
+  if (params->flags & IREE_UK_FLAG_MMT4D_ACCUMULATE) {
+    for (int i = 0; i < 4 * M0; ++i) {
+      acc[i] = vld1q_s32(out_ptr + 4 * i);
+    }
+  } else {
+    for (int i = 0; i < 4 * M0; ++i) {
+      acc[i] = vdupq_n_s32(0);
+    }
+  }
+  const int8x8_t vmask0f = vmov_n_s8(INT8_C(0x0F));
+  for (int k = 0; k < params->K; ++k) {
+    // Handle 1x16x2.
+    int16_t lhs_0 = *lhs_ptr++;
+    int16_t lhs_1 = *lhs_ptr++;
+
+    int8x8_t rhs_0 = vld1_s8(rhs_ptr);
+    rhs_ptr += 8;
+    int8x8_t rhs_1 = vld1_s8(rhs_ptr);
+    rhs_ptr += 8;
+
+    int16x8_t rhs_00 = vmovl_s8(vand_s8(rhs_0, vmask0f));
+    int16x8_t rhs_10 = vmovl_s8(vshr_n_s8(rhs_0, 4));
+    int16x8_t rhs_01 = vmovl_s8(vand_s8(rhs_1, vmask0f));
+    int16x8_t rhs_11 = vmovl_s8(vshr_n_s8(rhs_1, 4));
+
+    acc[0] = vmlal_n_s16(acc[0], vget_low_s16(rhs_00), lhs_0);
+    acc[1] = vmlal_n_s16(acc[1], vget_high_s16(rhs_00), lhs_0);
+    acc[2] = vmlal_n_s16(acc[2], vget_low_s16(rhs_01), lhs_0);
+    acc[3] = vmlal_n_s16(acc[3], vget_high_s16(rhs_01), lhs_0);
+
+    acc[0] = vmlal_n_s16(acc[0], vget_low_s16(rhs_10), lhs_1);
+    acc[1] = vmlal_n_s16(acc[1], vget_high_s16(rhs_10), lhs_1);
+    acc[2] = vmlal_n_s16(acc[2], vget_low_s16(rhs_11), lhs_1);
+    acc[3] = vmlal_n_s16(acc[3], vget_high_s16(rhs_11), lhs_1);
+
+    if (M0 >= 2) {
+      // Handle 2x16x2.
+      int16_t lhs_2 = *lhs_ptr++;
+      int16_t lhs_3 = *lhs_ptr++;
+
+      acc[4] = vmlal_n_s16(acc[4], vget_low_s16(rhs_00), lhs_2);
+      acc[5] = vmlal_n_s16(acc[5], vget_high_s16(rhs_00), lhs_2);
+      acc[6] = vmlal_n_s16(acc[6], vget_low_s16(rhs_01), lhs_2);
+      acc[7] = vmlal_n_s16(acc[7], vget_high_s16(rhs_01), lhs_2);
+
+      acc[4] = vmlal_n_s16(acc[4], vget_low_s16(rhs_10), lhs_3);
+      acc[5] = vmlal_n_s16(acc[5], vget_high_s16(rhs_10), lhs_3);
+      acc[6] = vmlal_n_s16(acc[6], vget_low_s16(rhs_11), lhs_3);
+      acc[7] = vmlal_n_s16(acc[7], vget_high_s16(rhs_11), lhs_3);
+    }
+
+    if (M0 == 4) {
```

Comment:
Can you find a way to make the source code for this function more concise and less repetitive, by doing some small-iteration-count for loops and putting some variables in arrays to iterate over (e.g. `lhs_{x}` ==> `lhs[x]`) ?  In the past we refrained from doing so to avoid performance issues with LLVM code generation, but that problem isn't current anymore.

Reply:
I made `lhs` a vector instead and it now loads 8 elements in the beginning of the kernel. I couldn't make any smaller-iteration-count for loops though since `vmlal_lane_s16` requires a constant for its third argument so calculating indexes like `i * 2 + 1` is not possible.

Reply:
Random question: what is the difference between vmlal and smlal? The code here seems to be doing pretty much what LLVM generates but LLVM emits pairs like the following:
```
    14ec: 4f482334      smlal2  v20.4s, v25.8h, v8.h[0]
    14f4: 0f482330      smlal   v16.4s, v25.4h, v8.h[0]
```
which are basically multiplying and accumulating 4x16-bit high/low sub-vectors into a 4x32-bit accumulator vectors. Is there any reason to use one over the other?

Reply:
@dcaballe `smlal` is the instruction name in A64 assembly. `vmlal` is the intrinsic name, that compiles to `smlal` in A64. It's named after the earlier armv7/A32 instruction.